userspace/libsinsp/tracer_emitter.cpp

```diff
@@ -134,50 +134,74 @@ void tracer_writer::close_fd()
 	}
 }
 
-tracer_emitter::tracer_emitter(std::string tag, uint64_t timeout_ns)
+tracer_emitter::tracer_emitter(std::string tag, uint64_t timeout_ns, bool delay_enter)
 	: m_tag(std::move(tag))
-	, m_start_ns(sinsp_utils::get_current_time_ns())
+	, m_start_ns(m_enabled ? sinsp_utils::get_current_time_ns() : 0)
 	, m_timeout_ns(timeout_ns)
+	, m_delay_enter(delay_enter)
 {
-	start();
+	if (!m_delay_enter)
+	{
+		write_tracer(true);
+	}
 }
 
 bool tracer_emitter::m_enabled = false;
 
 // XXX find/write a constexpr-compatible string class
 // for compile time concatenation
-tracer_emitter::tracer_emitter(std::string tag, const tracer_emitter &parent, uint64_t timeout_ns)
+tracer_emitter::tracer_emitter(std::string tag, const tracer_emitter &parent,
+			       uint64_t timeout_ns, bool delay_enter)
 	: tracer_emitter::tracer_emitter(
-		parent.tag() + '.' + std::move(tag),
-		std::min(timeout_ns, parent.m_timeout_ns))
+		parent.m_tag + '.' + std::move(tag),
+		std::min(timeout_ns, parent.m_timeout_ns),
+		parent.m_delay_enter || delay_enter)
 {
 }
 
 tracer_emitter::~tracer_emitter()
 {
-	if (!m_exit_written)
-	{
-		write_tracer(false);
-		elapsed_time(); // just for the side effect of logging if needed
-	}
+	do_stop();
 }
 
-void tracer_emitter::start()
+uint64_t tracer_emitter::stop()
 {
-	write_tracer(true);
+	ASSERT(!m_exit_done);
+	return do_stop();
 }
 
-uint64_t tracer_emitter::stop()
+uint64_t tracer_emitter::do_stop()
 {
-	ASSERT(!m_exit_written);
-	if (!m_exit_written)
+	if (!m_enabled || m_exit_done)
```

Comment:
Nit: you can avoid the `m_exit_done = true;` duplication by doing:

    m_exit_done = true;
    if (!m_enabled || m_exit_done)
    {
            return 0;
    }